xbmc/utils/StringUtils.cpp

```diff
@@ -327,96 +203,422 @@ std::wstring StringUtils::FormatV(const wchar_t *fmt, va_list args)
+  return utf8Str;
+}
+
+std::string StringUtils::ToUtf8(const std::wstring_view str)
 {
-  if (*(const wchar_t*)a <  *(const wchar_t*)b)
+  std::string utf8Str = CUnicodeConverter::WToUtf8(str, FAIL_ON_ERROR, SUBSTITUTE_BAD_CHARS);
+  return utf8Str;
+}
+
+std::u32string StringUtils::ToUtf32(const std::string_view str)
+{
+  std::u32string utf32Str =
+      CUnicodeConverter::Utf8ToUtf32(str, FAIL_ON_ERROR, SUBSTITUTE_BAD_CHARS);
+  return utf32Str;
+}
+
+std::u32string StringUtils::ToUtf32(const std::wstring_view str)
+{
+  std::u32string utf32Str;
+  utf32Str = CUnicodeConverter::WToUtf32(str, FAIL_ON_ERROR, SUBSTITUTE_BAD_CHARS);
+  return utf32Str;
+}
+
+std::wstring StringUtils::ToWstring(const std::string_view str)
+{
+  std::wstring wStr = CUnicodeConverter::Utf8ToW(str, FAIL_ON_ERROR, SUBSTITUTE_BAD_CHARS);
+  return wStr;
+}
+
+std::wstring StringUtils::ToWstring(const std::u32string_view str)
+{
+  std::wstring result;
+  result = CUnicodeConverter::Utf32ToW(str, FAIL_ON_ERROR, SUBSTITUTE_BAD_CHARS);
+  return result;
+}
+
+namespace
+{
+static int FoldAndCompare(const std::u32string_view str1, const std::u32string_view str2)
+{
+  // FoldAndEquals is faster for equality checks.
+
+  size_t length = std::min(str1.length(), str2.length());
+  for (size_t i = 0; i < length; i++)
+  {
+    char32_t fold_c1 = FoldCaseChar(str1[i]);
+    char32_t fold_c2 = FoldCaseChar(str2[i]);
+
+    if (fold_c1 != fold_c2)
+    {
+      if (fold_c1 < fold_c2)
+        return -1;
+      else
+        return 1;
+    }
+  }
+  // Break a tie using length
+
+  if (str1.length() < str2.length())
     return -1;
-  else if (*(const wchar_t*)a >  *(const wchar_t*)b)
+  if (str2.length() < str1.length())
     return 1;
+
   return 0;
 }
+} // namespace
 
-wchar_t tolowerUnicode(const wchar_t& c)
+std::locale StringUtils::GetCLocale()
 {
-  wchar_t* p = (wchar_t*) bsearch (&c, unicode_uppers, sizeof(unicode_uppers) / sizeof(wchar_t), sizeof(wchar_t), compareWchar);
-  if (p)
-    return *(unicode_lowers + (p - unicode_uppers));
+  static std::locale C_LOCALE;
+  static bool initialized{false};
 
-  return c;
+  if (!initialized)
+  {
+    std::locale x;
+    try
+    {
+      x = std::locale("C.UTF-8"); // MacOS does not like C.UTF-8
+      C_LOCALE = x;
+      initialized = true;
+    }
+    catch (std::runtime_error& e)
+    {
+      // Have to be careful about logging in StringUtils. Can get recursive during
+      // formatting of message.
+
+      std::cout << "Locale C.UTF-8 not supported trying 'C'" << std::endl;
+      try
+      {
+        x = std::locale("C"); // This shouldn't fail
+        C_LOCALE = x;
+        initialized = true;
+      }
+      catch (std::runtime_error& e2)
+      {
+        // Have to be careful about logging in StringUtils. Can get recursive
+
+        std::cout << "Locale C not supported" << std::endl;
+      }
+    }
+  }
+  return C_LOCALE;
```

Comment:
Isn't this prone to data races? I think the best option to synchronize this is to initialize the `std::locale` in a lambda like this:
```c++
static const std::locale C_LOCALE = []() {
  // everything from the "if (!initialized)" block
}();
```
Not the `()` at the end to directly invoke the lambda. This also solves the problem that in case of a failure the initialization is repeated over and over again.